File: src/TradingSystem.Web/Services/AuthService.cs
using System.Net.Http.Json;
using Microsoft.JSInterop;
using TradingSystem.Web.Models;

namespace TradingSystem.Web.Services;

public interface IAuthService
{
    Task<LoginResponse> LoginAsync(LoginRequest request);
    Task<RegisterResponse> RegisterAsync(RegisterRequest request);
    Task<bool> RefreshTokenAsync();
    Task LogoutAsync();
    Task<string?> GetTokenAsync();
    Task<string?> GetUsernameAsync();
    Task<bool> IsAuthenticatedAsync();
    Task<SendOtpResponse> SendOtpAsync(string email);
    Task<VerifyOtpResponse> VerifyOtpAsync(string email, string code);
    Task<ChangePasswordResponse> ChangePasswordAsync(ChangePasswordRequest request);
    Task<bool> CheckUsernameExistsAsync(string username);
}

public class AuthService : IAuthService
{
    private readonly HttpClient _httpClient;
    private readonly IJSRuntime _jsRuntime;
    private const string TokenKey = "authToken";
    private const string UserKey = "authUser";

    public AuthService(HttpClient httpClient, IJSRuntime jsRuntime)
    {
        _httpClient = httpClient;
        _jsRuntime = jsRuntime;
    }

    public async Task<LoginResponse> LoginAsync(LoginRequest request)
    {
        try
        {
            var response = await _httpClient.PostAsJsonAsync("/auth/login", request);
            var result = await response.Content.ReadFromJsonAsync<LoginResponse>();

            if (result == null)
                return new LoginResponse { Success = false, Error = "Invalid response from server" };

            if (result.Success && !string.IsNullOrEmpty(result.Token))
            {
                await _jsRuntime.InvokeVoidAsync("localStorage.setItem", TokenKey, result.Token);
                await _jsRuntime.InvokeVoidAsync("localStorage.setItem", UserKey,
                    System.Text.Json.JsonSerializer.Serialize(result.User));
            }

            return result;
        }
        catch (Exception ex)
        {
            return new LoginResponse { Success = false, Error = $"Connection failed: {ex.Message}" };
        }
    }

    public async Task<RegisterResponse> RegisterAsync(RegisterRequest request)
    {
        try
        {
            var response = await _httpClient.PostAsJsonAsync("/auth/register", request);
            var result = await response.Content.ReadFromJsonAsync<RegisterResponse>();

            return result ?? new RegisterResponse { Success = false, Error = "Invalid response from server" };
        }
        catch (Exception ex)
        {
            return new RegisterResponse { Success = false, Error = $"Connection failed: {ex.Message}" };
        }
    }

    public async Task LogoutAsync()
    {
        await _jsRuntime.InvokeVoidAsync("localStorage.removeItem", TokenKey);
        await _jsRuntime.InvokeVoidAsync("localStorage.removeItem", UserKey);
    }

    public async Task<string?> GetTokenAsync()
    {
        try
        {
            return await _jsRuntime.InvokeAsync<string?>("localStorage.getItem", TokenKey);
        }
        catch
        {
            return null;
        }
    }

    public async Task<bool> IsAuthenticatedAsync()
    {
        var token = await GetTokenAsync();
        return !string.IsNullOrEmpty(token);
    }

    public async Task<string?> GetUsernameAsync()
    {
        try
        {
            var userJson = await _jsRuntime.InvokeAsync<string?>("localStorage.getItem", UserKey);
            if (!string.IsNullOrEmpty(userJson))
            {
                var user = System.Text.Json.JsonSerializer.Deserialize<TradingSystem.Web.Models.UserInfo>(userJson,
                    new System.Text.Json.JsonSerializerOptions { PropertyNameCaseInsensitive = true });
                return user?.Username;
            }
        }
        catch { }
        return null;
    }

    public async Task<bool> RefreshTokenAsync()
    {
        try
        {
            var username = await GetUsernameAsync();
            if (string.IsNullOrEmpty(username)) return false;

            var response = await _httpClient.PostAsJsonAsync("/auth/refresh", new { Username = username });
            var result = await response.Content.ReadFromJsonAsync<LoginResponse>();

            if (result?.Success == true && !string.IsNullOrEmpty(result.Token))
            {
                await _jsRuntime.InvokeVoidAsync("localStorage.setItem", TokenKey, result.Token);
                await _jsRuntime.InvokeVoidAsync("localStorage.setItem", UserKey,
                    System.Text.Json.JsonSerializer.Serialize(result.User));
                return true;
            }
            return false;
        }
        catch
        {
            return false;
        }
    }

    public async Task<SendOtpResponse> SendOtpAsync(string email)
    {
        try
        {
            var response = await _httpClient.PostAsJsonAsync("/auth/send-otp", new { Email = email });
            var result = await response.Content.ReadFromJsonAsync<SendOtpResponse>();

            return result ?? new SendOtpResponse { Success = false, Error = "Invalid response from server" };
        }
        catch (Exception ex)
        {
            return new SendOtpResponse { Success = false, Error = $"Connection failed: {ex.Message}" };
        }
    }

    public async Task<VerifyOtpResponse> VerifyOtpAsync(string email, string code)
    {
        try
        {
            var response = await _httpClient.PostAsJsonAsync("/auth/verify-otp", new { Email = email, Code = code });
            var result = await response.Content.ReadFromJsonAsync<VerifyOtpResponse>();

            return result ?? new VerifyOtpResponse { Success = false, Error = "Invalid response from server" };
        }
        catch (Exception ex)
        {
            return new VerifyOtpResponse { Success = false, Error = $"Connection failed: {ex.Message}" };
        }
    }

    public async Task<ChangePasswordResponse> ChangePasswordAsync(ChangePasswordRequest request)
    {
        try
        {
            var response = await _httpClient.PostAsJsonAsync("/auth/change-password", request);
            var result = await response.Content.ReadFromJsonAsync<ChangePasswordResponse>();
            return result ?? new ChangePasswordResponse { Success = false, Error = "Invalid response from server" };
        }
        catch (Exception ex)
        {
            return new ChangePasswordResponse { Success = false, Error = $"Connection failed: {ex.Message}" };
        }
    }

    public async Task<bool> CheckUsernameExistsAsync(string username)
    {
        try
        {
            var response = await _httpClient.GetAsync($"/auth/check-username/{Uri.EscapeDataString(username)}");
            if (response.IsSuccessStatusCode)
            {
                var result = await response.Content.ReadFromJsonAsync<UsernameCheckResponse>();
                return result?.Exists ?? false;
            }
            return false;
        }
        catch
        {
            return false;
        }
    }

    private class UsernameCheckResponse
    {
        public bool Exists { get; set; }
    }
}

File: src/TradingSystem.Angular/src/app/core/services/auth.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { BehaviorSubject, Observable, of } from 'rxjs';
import { tap, catchError, map } from 'rxjs/operators';
import {
  LoginRequest, LoginResponse, RegisterRequest, RegisterResponse,
  UserInfo, SendOtpResponse, VerifyOtpResponse,
  ChangePasswordRequest, ChangePasswordResponse
} from '../../models/models';

@Injectable({ providedIn: 'root' })
export class AuthService {
  private readonly TOKEN_KEY = 'authToken';
  private readonly USER_KEY = 'authUser';

  private userSubject = new BehaviorSubject<UserInfo | null>(this.getStoredUser());
  user$ = this.userSubject.asObservable();

  constructor(private http: HttpClient) {}

  login(request: LoginRequest): Observable<LoginResponse> {
    return this.http.post<LoginResponse>('/auth/login', request).pipe(
      tap(result => {
        if (result.success && result.token) {
          localStorage.setItem(this.TOKEN_KEY, result.token);
          localStorage.setItem(this.USER_KEY, JSON.stringify(result.user));
          this.userSubject.next(result.user || null);
        }
      }),
      catchError(err => of({ success: false, error: `Connection failed: ${err.message}` } as LoginResponse))
    );
  }

  register(request: RegisterRequest): Observable<RegisterResponse> {
    return this.http.post<RegisterResponse>('/auth/register', request).pipe(
      catchError(err => of({ success: false, error: `Connection failed: ${err.message}` } as RegisterResponse))
    );
  }

  logout(): void {
    localStorage.removeItem(this.TOKEN_KEY);
    localStorage.removeItem(this.USER_KEY);
    this.userSubject.next(null);
  }

  refreshToken(): Observable<boolean> {
    const username = this.getUsername();
    if (!username) return of(false);
    return this.http.post<LoginResponse>('/auth/refresh', { username }).pipe(
      map(result => {
        if (result?.success && result.token) {
          localStorage.setItem(this.TOKEN_KEY, result.token);
          localStorage.setItem(this.USER_KEY, JSON.stringify(result.user));
          this.userSubject.next(result.user || null);
          return true;
        }
        return false;
      }),
      catchError(() => of(false))
    );
  }

  getToken(): string | null {
    return localStorage.getItem(this.TOKEN_KEY);
  }

  getUsername(): string | null {
    const user = this.getStoredUser();
    return user?.username || null;
  }

  isAuthenticated(): boolean {
    return !!this.getToken();
  }

  getUser(): UserInfo | null {
    return this.userSubject.value;
  }

  sendOtp(email: string): Observable<SendOtpResponse> {
    return this.http.post<SendOtpResponse>('/auth/send-otp', { email }).pipe(
      catchError(err => of({ success: false, error: `Connection failed: ${err.message}` } as SendOtpResponse))
    );
  }

  verifyOtp(email: string, code: string): Observable<VerifyOtpResponse> {
    return this.http.post<VerifyOtpResponse>('/auth/verify-otp', { email, code }).pipe(
      catchError(err => of({ success: false, error: `Connection failed: ${err.message}` } as VerifyOtpResponse))
    );
  }

  changePassword(request: ChangePasswordRequest): Observable<ChangePasswordResponse> {
    return this.http.post<ChangePasswordResponse>('/auth/change-password', request).pipe(
      catchError(err => of({ success: false, error: `Connection failed: ${err.message}` } as ChangePasswordResponse))
    );
  }

  checkUsernameExists(username: string): Observable<boolean> {
    return this.http.get<{ exists: boolean }>(`/auth/check-username/${encodeURIComponent(username)}`).pipe(
      map(r => r?.exists ?? false),
      catchError(() => of(false))
    );
  }

  sendPhoneOtp(phoneNumber: string, countryCode: string): Observable<SendOtpResponse> {
    return this.http.post<SendOtpResponse>('/auth/send-phone-otp', { phoneNumber, countryCode }).pipe(
      catchError(err => of({ success: false, error: err.message } as SendOtpResponse))
    );
  }

  verifyPhoneOtp(phoneNumber: string, countryCode: string, code: string): Observable<VerifyOtpResponse> {
    return this.http.post<VerifyOtpResponse>('/auth/verify-phone-otp', { phoneNumber, countryCode, code }).pipe(
      catchError(err => of({ success: false, error: err.message } as VerifyOtpResponse))
    );
  }

  private getStoredUser(): UserInfo | null {
    try {
      const json = localStorage.getItem(this.USER_KEY);
      return json ? JSON.parse(json) : null;
    } catch {
      return null;
    }
  }
}
